samza-core/src/main/java/org/apache/samza/operators/spec/SinkOperatorSpec.java

```diff
@@ -44,45 +50,36 @@
   private final int opId;
 
   /**
-   * The user-defined sink function
+   * The sink function to send messages
    */
   private final SinkFunction<M> sinkFn;
 
-  /**
-   * Potential output stream defined by the {@link SinkFunction}
-   */
-  private final OutputStream<M> outStream;
 
   /**
-   * Default constructor for a {@link SinkOperatorSpec} w/o an output stream. (e.g. output is sent to remote database)
+   * Constructs a {@link SinkOperatorSpec} with a user defined {@link SinkFunction}.
    *
    * @param sinkFn  a user defined {@link SinkFunction} that will be called with the output message,
    *                the output {@link org.apache.samza.task.MessageCollector} and the
    *                {@link org.apache.samza.task.TaskCoordinator}.
-   * @param opCode  the specific {@link OpCode} for this {@link SinkOperatorSpec}. It could be {@link OpCode#SINK}, {@link OpCode#SEND_TO},
-   *                or {@link OpCode#PARTITION_BY}
-   * @param opId  the unique id of this {@link SinkOperatorSpec} in the {@link org.apache.samza.operators.StreamGraph}
+   * @param opCode  the specific {@link OpCode} for this {@link SinkOperatorSpec}.
+   *                It could be {@link OpCode#SINK}, {@link OpCode#SEND_TO}, or {@link OpCode#PARTITION_BY}
+   * @param opId  the unique ID of this {@link SinkOperatorSpec} in the {@link org.apache.samza.operators.StreamGraph}
```

Comment:
SAMZA-1101 is included in this patch. You're right, the idea is to create a different MessageCollector (or an even lower abstraction like the SystemConsumer/Producer) instead of creating a different SinkFunction for test vs prod environments. The user-provided or generated SinkFunction only describes the transforms for the messages and identity of the stream before sending it out via the collector. The environment abstracts how the collector functions (e.g. sending to a Kafka stream vs adding to an in-memory collection).